### PR TITLE
feat(gazelle): Add "python_default_visibility" directive

### LINKED ISSUE #1682: gazelle: Visibilities are incorrect for projects with multiple `# gazelle:python_root` directives
# 🐞 bug report

### Affected Rule

 The issue is caused by the rule: https://github.com/bazelbuild/rules_python/blob/b106f91c9da7e31d73a7293e88ac78eedcad2057/gazelle/python/generate.go#L215

### Is this a regression?

 No

### Description

This is necessary in monorepo situations where there are multiple `# gazelle:python_root` directives used, e.g. in the structure below:

```
ROOT
    proj_1
        src  # is a python root and depends on proj_2
            foo
    proj_2
        src  # is a python root
            bar  ← The visibility here will be incorrect; proj_1 cannot see this.
```

As it stands, we write a narrowly scoped `visibility` attribute for every generated rule; the visibility is `//{python-root-package}:__subpackages__`. Bazel will __fail__ to build targets in `proj_1` because of this.

## 🔥 Exception or Error

<pre><code>

</code></pre>

## 🌍  Your Environment

**Operating System:**

<pre>
  <code>
macOS 14.2
  </code>
</pre>

**Output of `bazel version`:**

<pre>
  <code>
Build label: 7.0.0
Build target: @@//src/main/java/com/google/devtools/build/lib/bazel:BazelServer
Build time: Mon Dec 11 16:52:42 2023 (1702313562)
Build timestamp: 1702313562
Build timestamp as int: 1702313562
  </code>
</pre>

**Rules_python version:**

<pre>
  <code>
0.27.1
  </code>
</pre>

**Anything else relevant?**

The desired implementation is not to set everything to `//visibility:public`, rather, it's to resolve cross-project dependencies in a monorepo the same way as they are within one directory.

Consider a working example here: https://github.com/bazelbuild/rules_python/compare/main...aryamccarthy:rules_python:main

### LINKED ISSUE #1783: Add `python_visibility` gazelle directive
# 🚀 feature request

### Relevant Rules

+ gazelle

### Description

**Summary:** Add a `py(thon)_visibility` gazelle directive that acts like the `go_visibility` directive. This directive would append additional labels to the `visibility` attribute on generated targets.

**Example:**

```starlark
# gazelle:python_visibility //tests:__pkg__
# gazelle:python_visibility //bar:baz

py_library(
    name = "directive_python_visibility",
    srcs = ["foo.py"],
    visibility = [
        "//:__subpackages__",
        "//bar:baz",
        "//tests:__pkg__",
    ],
)
```

### Describe the solution you'd like

I have a branch that does this already. I'm mostly making this Issue for discussion before I submit the PR.

Branch: https://github.com/dougthor42/rules_python/tree/gazelle-python-visibility

### Describe alternatives you've considered

Manually adding other vis labels is the only other option I've found. That's not really an option when adding gazelle to an existing, large project.

### PR DESCRIPTION
**Pull request by dougthor42**
Fixes #1783. Provides a way to fix #1682.

Add the `python_default_visibility` directive. This directive sets the
`visibility` attribute on all generated `py_*` rules. It accepts a
comma-separated list of labels to add as visibility targets, similar
to how the base `default_visibility` directive works.

Setting this directive multiple times will override previous values.

Two special values are also accepted: `NONE` and `DEFAULT`. See
./gazelle/README.md#directive-python_default_visibility for details.

The directive also accepts a special string `"$python_root"` that gets
replaced with the `python_root` value, if set. If not set, `"$python_root"`
is replaced with the empty string.

### CODE REVIEW

**Comment by dougthor42** [comment]
I think this is ready for review now.

Sorry for the big PR! Most of it is just test cases though.

File: gazelle/python/configure.go [hunk 1]
```diff
@@ -163,6 +167,20 @@ func (py *Configurer) Configure(c *config.Config, rel string, f *rule.File) {
 			config.SetBinaryNamingConvention(strings.TrimSpace(d.Value))
 		case pythonconfig.TestNamingConvention:
 			config.SetTestNamingConvention(strings.TrimSpace(d.Value))
+		case pythonconfig.DefaultVisibilty:
+			switch directiveArg := strings.TrimSpace(d.Value); directiveArg {
+			case "NONE":
+				config.SetDefaultVisibility([]string{})
+			case "DEFAULT":
```
Comment (dougthor42) [inline gazelle/python/configure.go:174]: I also thought of `"RESET"` instead.

File: gazelle/python/configure.go [hunk 2]
```diff
@@ -99,6 +100,8 @@ func (py *Configurer) Configure(c *config.Config, rel string, f *rule.File) {
 	}
 
 	gazelleManifestFilename := "gazelle_python.yaml"
+	// TODO: figure out how to keep this in sync with pythonconfig.go New *Config
+	defaultVisibilityFmtString := "//%s:__subpackages__"
```
Comment (f0rmiga) [inline gazelle/python/configure.go:104]: Make this an exported `const` in the `pythonconfig.go`.

File: gazelle/python/configure.go (hunk 1 above)
Comment (f0rmiga) [reply gazelle/python/configure.go:174]: I think `DEFAULT` is more self-descriptive.

File: gazelle/README.md [hunk 3]
```diff
@@ -238,6 +240,84 @@ py_libary(
...
+    ...,
+)
+```
+
+You can also inject the `python_root` value by using the exact string
+`$python_root`:
+
+```starlark
+# an absurd example
+# gazelle:python_default_visibility //$python_root:__pkg__,//foo/$python_root/tests:__subpackages__,//$python_root/$python_root/foo:bar
```
Comment (f0rmiga) [inline gazelle/README.md:273]: Thinking more, `//$python_root/$python_root/foo:bar` will probably confuse users. Let's keep this more straightforward and remove it?

File: gazelle/python/testdata/directive_python_default_visibility/test6_default_label/subpkg/BUILD.out [hunk 4]
```diff
@@ -0,0 +1,10 @@
+load("@rules_python//python:defs.bzl", "py_library")
+
+# Reset the default visibility to the default for all child packages.
+# gazelle:python_default_visibility DEFAULT
```
Comment (f0rmiga) [inline gazelle/python/testdata/directive_python_default_visibility/test6_default_label/subpkg/BUILD.out:4]: What if we set `# gazelle:python_visibility` together with `# gazelle:python_default_visibility`?

Comment (f0rmiga) [reply gazelle/python/testdata/directive_python_default_visibility/test6_default_label/subpkg/BUILD.out:4]: How about a test case that uses `# gazelle:python_visibility` on a sub-package?

File: gazelle/README.md [hunk 5]
```diff
@@ -238,6 +240,84 @@ py_libary(
...
+
+# results in:
+py_library(
+    ...,
+    visibility = ["//:__subpackages__"],
+    ...,
+)
+```
+
+These special values can be useful for descendant `BAZEL.build` packages.
```
Comment (f0rmiga) [inline gazelle/README.md:318]:
```suggestion
These special values can be useful for sub-packages.
```

**Review by f0rmiga — CHANGES_REQUESTED** [review]

File: gazelle/python/configure.go (hunk 2 above)
Comment (dougthor42) [reply gazelle/python/configure.go]: Thanks, done.

**Review by aignas — COMMENTED** [review]
LGTM, but would be great to have a `CHANGELOG.md` entry announcing the feature. I think pointing users to `gazelle` docs would be sufficient enough.

It's really nice to have extensive test coverage!

@f0rmiga, what do you think about readiness of this?

**Review by f0rmiga — APPROVED** [review]

**Comment by f0rmiga** [comment]
Thanks for addressing my comments! Thanks @aignas for the review. I'll let you merge this when the changelog looks good to you.

**Review by aignas — APPROVED** [review]

**Pull request merged**